### DISCUSSION TITLE
A question about CPU vs Edge TPU

### DISCUSSION THREAD

**Discussion by MrAnonymous722** [blakeblackshear/frigate#1532] [Ask A Question]
Hello, 

I'm currently running Frigate on a Ryzen 7 3700x in a Virtualbox guest hosted on Windows 10. I'm seeing what I think to be fine performance with 2 coral detectors running on CPU (Not sure if this even makes it faster, I really don't know what I'm doing). 

I guess what I'm curious about is what should I look for as far as signs I should get a Coral device instead of just sticking with my CPU. I think really my only limiter to latency is from my Wyze RTSP bridge since Wyze won't make a native RTSP firmware for their V3 cams.

I know this is off topic to this project, but do you have any suggestions for alternatives to the USB Coral TPU since it seems to be out of stock or highly marked up. Would I be able to use an M-Key style on a desktop motherboard with an M.2 slot? Any feedback or suggestions is greatly appreciated!

**Comment by blakeblackshear** [comment]
What are your inference times in the debug page for frigate?

**Reply by MrAnonymous722** [reply]
Looking in the debug page, I guess I actually have 3. They're each around 300.

**Comment by blakeblackshear** [comment]
A Coral will do inference in less than 10ms on many machines, so that's 30x faster than your Ryzen. You probably won't be able to get those speeds on a windows host with VirtualBox in the middle though. Note that even a RPi3 with a Coral blows the performance of your machine out of the water for object detection. 

If it takes 300ms to run object detection, that means frigate can process at most 3fps, probably less because object detection is often run multiple times per frame. I would also run a detector per camera if you have the CPU to spare. 

In my opinion, there is no situation where it makes sense to not use a Coral. It's fine to use CPU while you wait for them to be available though. 

I would keep an eye out on the number of skipped frames. Definitely set the fps in your config to 5 or less even though the Wyzecams output higher fps.

**Reply by MrAnonymous722** [reply]
Well that definitely answers my question on CPU vs Coral haha. As far as a B+M Key Coral TPU fitting into a x470 motherboard with storage M.2 slots, do you have any suggestions? I understand that's not really within the scope here, so if not I understand. It seems like I can get one of these for MSRP. They even have dual TPU A+E key devices for under $40, but I am doubtful of that fitting with my motherboard.

**Reply by blakeblackshear** [reply]
I don't. I know others have gotten the other Corals to work with various adapters.

**Reply by MrAnonymous722** [reply]
Gotcha. So probably best just to wait till the USB one is back at a reasonable price. Thank you for everything!

**Reply by shred86** [reply]
The M.2 B+M key should fit perfectly into a M.2 storage slot. I’m pretty sure that’s exactly what it was designed for. I added one to my AWOL Mini PC that had an extra M.2 NVME storage slot and it fits just fine.

**Reply by MrAnonymous722** [reply]
And it works fine with Frigate? I might get one since I have an open slot and they're available for cheap now.

**Reply by shred86** [reply]
Yep, works great. I actually prefer an internal device over an external USB dongle... and it's cheaper.

**Reply by ozett** [reply]
> I would also run a detector per camera

is there a speical way to configure this? Or should i use as many cpus entries as there are cameras?
like 
```
3 cams ->

detectors:
  cpu1:
    type: cpu
  cpu2:
    type: cpu
  cpu3:
    type: cpu
```

and accordingly for 4,5,6....?

**Comment by lawm** [comment]
Just to give another data point,
I used to use frigate in a VM in Hyper-V on Windows and got 250ms CPU inference times.
Now, I use it in a docker container on Debian and get 130ms.

I only have 1 camera at 5fps, so it's ok for me.

**Reply by MrAnonymous722** [reply]
I absolutely intend on running it on a native Ubuntu host once I get all my stuff moved to my new house. Just wanted to tinker with it on my main machine until then. I am curious to see the better inference times in action though, because even 300 seems to be pretty quick for my couple of cameras. But I think I have a pretty ideal setup with minimal motion other than the objects I'm looking for.

**Comment by itkfilelor** [comment]
I was able to get a mini-PCIe on sale fro $15 and picked up a mini to PCIe 4x adapter for $10, less than the usb and both arrived in 3 days. work nice. my inference is about 10ms with 4 cameras currently.

For those unaware, mini-PCIe is the slot that usually houses your wifi and or bluetooth chips on a laptop, they are rare on a desktop these days.

**Reply by MrAnonymous722** [reply]
This will likely be the route I need to take. I had a little oversight and my server machine doesn’t have M.2 slots in it so I’d need to do this or the USB device. Would you happen to know if it’s possible to pass through one of these devices thru virtual box by chance haha

**Reply by MrAnonymous722** [reply]
Are you able to drop a link for that adapter you got? Or point me in the direction of a comparable one?
